### PR TITLE
TST/CI: Split Circle workflow

### PR DESCRIPTION
**Pull request by effigies**
Updating with todo list from comment thread:

* [x] Properly place codecov outputs, move codecov to its own entry in jobs (https://github.com/nipy/nipype/pull/2386#issuecomment-363870964)
* [x] Save coverage artifacts (https://github.com/nipy/nipype/pull/2386#issuecomment-363871713)
* [x] Save built docs as artifacts (https://github.com/nipy/nipype/pull/2386#issuecomment-363871713)
* [x] Test summaries (https://github.com/nipy/nipype/pull/2386#issuecomment-363872361)

----

#### Original text

This PR tracks an attempt to split out the testing process into a workflow of single-container jobs using the Circle 2.0 workflows. The goal is to reduce the amount of time containers are allocated but idle, and allow multiple PRs to do some simultaneous testing.

Follow up to #2202.

Comments and suggestions welcome. cc @kaczmarj

### CODE REVIEW

**Comment by effigies** [comment]
This is working. There's a definite slowdown due to saving and loading images, so perhaps builds should be moved back into the test jobs. Still, it looks like we can achieve concurrency across PRs.

**Comment by effigies** [comment]
From the [latest build](https://circleci.com/workflow-run/4f08c3bc-a38b-4500-bf67-de204ec74cbc):

17:58 to build images
15:14 to tarball images and persist to workspace

12:38 -12:47 to attach to workspace and load images

So for a cost of about 5 minutes per test job, we can skip a 30 minute bottleneck. Seems like a good trade-off.

File: .circleci/config.yml [hunk 1]
```diff
@@ -120,31 +153,159 @@ jobs:
...
+            bash /home/circleci/nipype/.circleci/test_py3_fmri_spm_dartel_linear_l2.sh
+      - store_artifacts:
+          path: /home/circleci/work/tests
+
+  test_fmri_spm_nested_multiproc:
+    docker:
+      - image: docker:17.10.0-ce-git
+    working_directory: /home/circleci/nipype
+    steps:
+      - setup_remote_docker
```
Comment (oesteban) [inline .circleci/config.yml]: the remote docker might be useful when building, but it is not necessary for running. I may be wrong, but I think that remote docker does not allow mounting, and you need to do `docker cp`. The different jobs may be just using the local docker install. IMHO this line can be removed on the test jobs.

File: .circleci/config.yml [hunk 2]
```diff
@@ -120,31 +153,159 @@ jobs:
...
+          command: |
+            docker load < /tmp/docker/nipype-base-latest-py36-py27.tar.gz
+      - checkout:
+          path: /home/circleci/nipype
+      - run:
+          name: Get codecov
+          command: |
+            pip install --no-cache-dir codecov
+      - run:
+          name: Download test data
```
Comment (oesteban) [inline .circleci/config.yml]: Maybe we can even split further and take this out as a different job?

File: .circleci/config.yml [hunk 3]
```diff
@@ -120,31 +153,159 @@ jobs:
...
-          no_output_timeout: 4h
+          no_output_timeout: 2h
           environment:
             WORKDIR: /home/circleci/work
           command: |
             mkdir -p "$WORKDIR"
             chmod -R 777 "$WORKDIR"
-            bash /home/circleci/nipype/.circleci/tests.sh
+            bash /home/circleci/nipype/.circleci/pytests.sh
       - store_artifacts:
```
Comment (oesteban) [inline .circleci/config.yml]: I'd use `store_test_results` here, so we can have the test summary.

**Comment by oesteban** [comment]
@kaczmarj could you have a quick look? Thanks!

**Comment by effigies** [comment]
There are some issues with codecov not pointing to the right files. I think I want to move the codecov to its own entry in Circle, which would let us simplify the script structure, too.

I just haven't found much time to figure out where the coverage files are/should be.

**Comment by satra** [comment]
also we should ensure that some of the artifacts are being saved, such as the built docs and the coverage files.

**Comment by oesteban** [comment]
oh, one more: use set up the test summaries (https://circleci.com/docs/2.0/configuration-reference/#store_test_results) so that circle can show us number of tests and time in the `test_build` job

**Comment by effigies** [comment]
Well, this branch is open to pushes from team members...

Anyway, I can try to find some time later this week, but I think it's going to take a couple periods of sustained attention, so no definite time frame.

**Comment by kaczmarj** [comment]
@effigies @oesteban - I will look soon. This is an exciting pr!

**Comment by kaczmarj** [comment]
Sorry for the delay. This looks great to me. One thing I would suggest is to reduce duplication in in `.circleci/config.yml` with yaml pointers. Common procedures can be defined once and referenced multiple times.

For example:

```yaml
_modify_nipype_version: &modify_nipype_version
  name: Modify Nipype version if necessary
  command: |
    if [ "$CIRCLE_TAG" != "" ]; then
      sed -i -E "s/(__version__ = )'[A-Za-z0-9.-]+'/\1'$CIRCLE_TAG'/" nipype/info.py
    fi
# ...
- run: *modify_nipype_version
# ...
- run: *modify_nipype_version
```

I can try to add this over the weekend. Otherwise this looks great!

**Comment by effigies** [comment]
That is super handy to know. If you just want to ping this thread when
you're going to work on something, that'll keep me from doing the same
thing if I happen to have time to work on it at the same time. (I'll do the
same.)

**Comment by kaczmarj** [comment]
@effigies - I am working on this now

**Comment by effigies** [comment]
This is ready for review. Docs and coverage are in the artifacts. Codecov is pushing pull request numbers again, which should start showing up again once this hits master. And test summaries show up for the `test_pytest` entry.

**Comment by satra** [comment]
thanks @effigies - this looks good.

**Pull request merged**